src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,215 @@
+			return {};
+		}
+		cores.push_back(current_core++);
+		++assigned_cores;
+	}
+	return cores;
+}
+
+struct pinned_thread_state {
+	cpu_set_t previous_cpuset = {};
+	cpu_set_t pinned_to_cpuset = {};
+};
+
+struct thread_pinner_state {
+	std::mutex mutex;
+	bool initialized = false;
+	runtime_configuration config;
+	cpu_set_t available_cores = {};
+	std::unordered_map<thread_type, uint32_t> thread_pinning_plan;
+	std::unordered_map<pthread_t, pinned_thread_state> pinned_threads;
+};
+thread_pinner_state g_state; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// This helper removes a thread from the pinned_threads map when it ends
+struct thread_remover {
+	thread_remover() = default;
+	thread_remover(const thread_remover&) = delete;
+	thread_remover& operator=(const thread_remover&) = delete;
+	thread_remover(thread_remover&&) = delete;
+	thread_remover& operator=(thread_remover&&) = delete;
+	~thread_remover() {
+		std::lock_guard lock(g_state.mutex);
+		if(g_state.initialized) { g_state.pinned_threads.erase(pthread_self()); }
 	}
+};
+thread_local std::optional<thread_remover> t_remover; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// Initializes the thread pinning machinery
+// This captures the current thread's affinity mask and sets the thread pinning machinery up
+// Calls to pin_this_thread prior to this call will have no effect
+void initialize(const runtime_configuration& cfg) {
+	std::lock_guard lock(g_state.mutex);
+	assert(!g_state.initialized && "Thread pinning already initialized.");
+	assert(g_state.thread_pinning_plan.empty() && "Thread pinning plan not initially empty.");
+	assert(g_state.pinned_threads.empty() && "Pinned threads not initially empty.");
+
+	g_state.config = cfg;
+
+	const auto ret = sched_getaffinity(0, sizeof(cpu_set_t), &g_state.available_cores);
+	if(ret != 0) {
+		CELERITY_WARN("Error retrieving initial process affinity mask, will disable pinning.");
+		g_state.config.enabled = false;
+		return;
+	}
+
+	// pinned threads per process: user, scheduler, executor, 1 backend worker per device if enabled
+	uint32_t pinned_threads_per_process = 3;
+	if(g_state.config.use_backend_device_submission_threads) { pinned_threads_per_process += g_state.config.num_devices; }
+	// total number of threads to be pinned across processes (legacy mode)
+	const uint32_t total_threads = pinned_threads_per_process * g_state.config.num_legacy_processes;
+
+	if(g_state.config.enabled) {
+		// select the core set to use
+		std::vector<uint32_t> selected_core_ids = {};
+		if(!cfg.hardcoded_core_ids.empty()) {
+			// just use the provided hardcoded IDs
+			if(static_cast<uint32_t>(cfg.hardcoded_core_ids.size()) != total_threads) {
+				CELERITY_WARN("Hardcoded core ID count ({}) does not match the number of threads to be pinned ({}), will disable pinning.",
+				    cfg.hardcoded_core_ids.size(), total_threads);
+			} else {
+				selected_core_ids = cfg.hardcoded_core_ids;
```

Comment:
Don't we also have to check whether the specific cores are even available to the process?

Reply:
It will fail later when pinning so I didn't want to add extra code here for this edge case.

Reply:
I just tried `CELERITY_THREAD_PINNING="4,5,6,7,8,9,10" taskset -c 49 ./examples/matmul/matmul` and didn't get any errors. In fact I think it simply ignored the process mask. Is that working as intended?

Reply:
Absolutely not intended, thanks for checking. I wrongly assumed that the `pthread_setaffinity_np` call would fail.

The behaviour for your example is now this:
```
[2024-11-21 11:28:59.479] [0] [warning] Not all hardcoded core IDs are available, downgrading to auto-pinning.
[2024-11-21 11:28:59.479] [0] [warning] Insufficient available cores for thread pinning (required 7, 1 available), disabling pinning.
```
There's also a new test for it.